File: examples/python/diagnostic_codes.py
"""
Diagnostic and coupling codes — opt-in (OFF by default).
Codes: D1, D3, D4, D5, D6, M2

These do not create false positives but hurt observability and maintainability.
Enable with `severity = { D1 = "info" }` in `.falsegreen.toml` or
`[tool.falsegreen]` in `pyproject.toml`.
"""
import pytest


# ─── D1: Assertion Roulette (2+ asserts, none with a message) ─────────────────

# BAD: when any assertion fails, the output says only the line number
def test_d1_assertion_roulette():
    order = build_order(items)
    assert subtotal(order) == 30    # D1 — which of the three failed?
    assert discount(order) == 3     # D1
    assert total(order) == 27       # D1

# CLEAN: at least one assertion has a message, or all do
def test_d1_with_messages():
    order = build_order(items)
    assert subtotal(order) == 30, "subtotal mismatch"
    assert discount(order) == 3,  "discount mismatch"
    assert total(order) == 27,    "total mismatch"

# CLEAN: single assertion — not Assertion Roulette
def test_d1_single_assert():
    assert total(build_order(items)) == 27


# ─── D3: Duplicate Assert (same assertion written twice) ─────────────────────

# BAD: the second assertion adds no coverage
def test_d3_duplicate_assert():
    result = compute()
    assert result == 42    # D3 — first occurrence
    assert result == 42    # D3 — exact duplicate

# BAD: identical assertion in different places
def test_d3_duplicate_in_branch():
    result = compute()
    if condition:
        assert result == 42
    assert result == 42    # D3 — also asserts outside the branch

# CLEAN: each assertion checks something distinct
def test_d3_clean():
    result = compute()
    assert result == 42
    assert isinstance(result, int)


# ─── D4: Unnamed Parametrize (3+ cases, no ids=) ────────────────────────────

# BAD: CI output shows test[0], test[1], test[2] — hard to debug
@pytest.mark.parametrize("value,expected", [
    ("alice", "ALICE"),
    ("bob",   "BOB"),
    ("carol", "CAROL"),     # D4 — 3+ cases, no ids=
])
def test_d4_unnamed(value, expected):
    assert value.upper() == expected

# CLEAN: human-readable IDs
@pytest.mark.parametrize("value,expected", [
    ("alice", "ALICE"),
    ("bob",   "BOB"),
    ("carol", "CAROL"),
], ids=["lowercase-name", "short-name", "another-name"])
def test_d4_with_ids(value, expected):
    assert value.upper() == expected


# ─── D5: Inline Setup Excess (too many setup statements before first assert) ──

# BAD: more than ~5 assignment/call statements before the first assert
def test_d5_excessive_setup():
    db = create_test_db()           # setup 1
    schema = load_schema("v2")      # setup 2
    user = create_user("Alice")     # setup 3
    session = db.open_session()     # setup 4
    session.save(user)              # setup 5
    role = Role("admin")            # setup 6
    user.assign_role(role)          # setup 7 — D5: too many inline setup lines
    assert user.has_role("admin")

# CLEAN: move setup to a fixture
@pytest.fixture
def admin_user(test_db):
    user = create_user("Alice")
    test_db.save(user)
    user.assign_role(Role("admin"))
    return user

def test_d5_clean(admin_user):
    assert admin_user.has_role("admin")


# ─── D6: Debug Print in test body ────────────────────────────────────────────

# BAD: print left over from debugging session
def test_d6_debug_print():
    result = compute()
    print(f"DEBUG result={result}")  # D6 — suppressed by pytest by default
    assert result == 42

# BAD: pprint
def test_d6_pprint():
    import pprint
    data = fetch_data()
    pprint.pprint(data)              # D6
    assert data["status"] == "ok"

# CLEAN: remove the print or replace with a proper log
def test_d6_clean():
    result = compute()
    assert result == 42


# ─── M2: Long Test Method ────────────────────────────────────────────────────

# BAD: a test body of 50+ lines tries to verify too many concerns at once.
# This is a structural signal to split into focused single-concern tests.

# Example of what NOT to do (abbreviated for readability):
def test_m2_too_long():
    # 50+ lines of setup + assertions testing multiple unrelated behaviors
    user = create_user("Alice")
    assert user.name == "Alice"           # concern 1: name
    assert user.role == "guest"           # concern 2: role default
    db.save(user)
    loaded = db.load(user.id)
    assert loaded.name == "Alice"         # concern 3: round-trip
    token = auth.issue_token(user)
    assert token.user_id == user.id       # concern 4: token issue
    assert not token.is_expired()         # concern 5: expiry
    # ... 45 more lines ...

# CLEAN: one test per concern
def test_m2_user_name_clean():
    assert create_user("Alice").name == "Alice"

def test_m2_user_default_role_clean():
    assert create_user("Alice").role == "guest"

def test_m2_user_round_trip_clean():
    user = create_user("Alice")
    db.save(user)
    assert db.load(user.id).name == "Alice"

File: examples/typescript/diagnostic_codes.ts
/**
 * Diagnostic and coupling codes - opt-in (OFF by default).
 * Codes: D1, D3, D4, D6, D7, D8, M2, C37
 *
 * These are NOT false-green. The test still protects against regressions; the
 * codes flag observability and maintainability problems. They run only on a
 * diagnostic pass, never as part of the default false-green detection.
 *
 * Enable with `severity = { D1 = "info" }` in the falsegreen config.
 *
 * TypeScript, Jest/Vitest idioms. Referenced symbols are illustrative.
 */

import { describe, it, test, expect } from 'vitest';
import { buildOrder, compute, fetchData } from './mymodule';

// ─── D1: Assertion Roulette (2+ asserts, none with a message) ────────────────

// BAD: when one of these fails, the output names only a line, not which check.
it('BAD d1 assertion roulette', () => {
  const order = buildOrder(items);
  expect(subtotal(order)).toBe(30); // D1 - which of the three failed?
  expect(discount(order)).toBe(3);  // D1
  expect(total(order)).toBe(27);    // D1
});

// CLEAN: each assertion carries a message
it('CLEAN d1 with messages', () => {
  const order = buildOrder(items);
  expect(subtotal(order), 'subtotal').toBe(30);
  expect(discount(order), 'discount').toBe(3);
  expect(total(order), 'total').toBe(27);
});


// ─── D3: Duplicate Assert (same assertion written twice) ─────────────────────

// BAD: the second assertion adds no coverage
it('BAD d3 duplicate assert', () => {
  const result = compute();
  expect(result).toBe(42); // D3 - first occurrence
  expect(result).toBe(42); // D3 - exact duplicate
});

// CLEAN: each assertion checks something distinct
it('CLEAN d3', () => {
  const result = compute();
  expect(result).toBe(42);
  expect(typeof result).toBe('number');
});


// ─── D4: Untitled it.each cases (no %/$ placeholder in the title) ─────────────

// BAD: a static title with no placeholder - every generated case shares the same
// name, so CI cannot tell them apart on failure.
it.each([
  ['alice', 'ALICE'],
  ['bob', 'BOB'],
  ['carol', 'CAROL'], // D4 - title has no placeholder, cases are indistinguishable
])('BAD d4 uppercases the input', (value, expected) => {
  expect(value.toUpperCase()).toBe(expected);
});

// CLEAN: a title with a placeholder gives each case a distinct, readable name
it.each([
  ['alice', 'ALICE'],
  ['bob', 'BOB'],
  ['carol', 'CAROL'],
])('CLEAN d4 uppercases %s to %s', (value, expected) => {
  expect(value.toUpperCase()).toBe(expected);
});


// ─── D6: console.* in a test body ─────────────────────────────────────────────

// BAD: a console.log left over from debugging clutters CI output
it('BAD d6 console.log', () => {
  const result = compute();
  console.log('DEBUG result=', result); // D6 - leftover debug output
  expect(result).toBe(42);
});

// BAD: console.debug is the same noise
it('BAD d6 console.debug', () => {
  const data = fetchData();
  console.debug(data); // D6
  expect(data.status).toBe('ok');
});

// CLEAN: no stray logging
it('CLEAN d6', () => {
  expect(compute()).toBe(42);
});


// ─── D7: Anonymous test (empty or missing description) ───────────────────────

// BAD: empty title - CI reports a blank test name
it('', () => {
  expect(compute()).toBe(42); // D7 - no description
});

// BAD: test() with only a callback and no name (where the runner allows it)
test(() => {
  expect(compute()).toBe(42); // D7 - anonymous
});

// CLEAN: a descriptive title
it('compute returns 42 for the default input', () => {
  expect(compute()).toBe(42);
});


// ─── D8: Magic number in an assertion ─────────────────────────────────────────

// BAD: 86400 and 30 carry meaning that the reader has to decode
it('BAD d8 magic numbers', () => {
  expect(secondsPerDay()).toBe(86400); // D8 - what is 86400?
  expect(retentionWindow()).toBe(2592000); // D8 - 30 days in seconds, unexplained
});

// CLEAN: name the constants so the assertion reads as intent
it('CLEAN d8 named constants', () => {
  const SECONDS_PER_DAY = 24 * 60 * 60;
  const RETENTION_DAYS = 30;
  expect(secondsPerDay()).toBe(SECONDS_PER_DAY);
  expect(retentionWindow()).toBe(RETENTION_DAYS * SECONDS_PER_DAY);
});


// ─── M2: Over-long test body ──────────────────────────────────────────────────

// BAD: one test verifies many unrelated concerns. A failure does not localize,
// and the body is hard to read. This is a structural signal to split it up.
it('BAD m2 too long', () => {
  const user = createUser('Alice');
  expect(user.name).toBe('Alice');        // concern 1: name
  expect(user.role).toBe('guest');        // concern 2: default role
  db.save(user);
  const loaded = db.load(user.id);
  expect(loaded.name).toBe('Alice');      // concern 3: round trip
  const token = auth.issueToken(user);
  expect(token.userId).toBe(user.id);     // concern 4: token issue
  expect(token.isExpired()).toBe(false);  // concern 5: expiry
  // ...many more concerns in the real case...
});

// CLEAN: one focused test per concern
it('CLEAN m2 name', () => {
  expect(createUser('Alice').name).toBe('Alice');
});
it('CLEAN m2 default role', () => {
  expect(createUser('Alice').role).toBe('guest');
});
it('CLEAN m2 round trip', () => {
  const user = createUser('Alice');
  db.save(user);
  expect(db.load(user.id).name).toBe('Alice');
});


// ─── C37: Duplicate it.each / test.each case ─────────────────────────────────

// BAD: (2, 3, 5) appears twice - the second run adds no coverage
it.each([
  [1, 1, 2],
  [2, 3, 5],
  [2, 3, 5], // C37 - exact duplicate of the line above
  [0, 0, 0],
])('BAD c37 add(%i, %i) === %i', (a, b, expected) => {
  expect(add(a, b)).toBe(expected);
});

// CLEAN: every case covers a distinct scenario
it.each([
  [1, 1, 2],
  [2, 3, 5],
  [-1, 1, 0],
  [0, 0, 0],
])('CLEAN c37 add(%i, %i) === %i', (a, b, expected) => {
  expect(add(a, b)).toBe(expected);
});


// ─── Placeholder stubs ────────────────────────────────────────────────────────

declare const items: unknown;
declare function subtotal(o: unknown): number;
declare function discount(o: unknown): number;
declare function total(o: unknown): number;
declare function secondsPerDay(): number;
declare function retentionWindow(): number;
declare function createUser(name: string): { id: number; name: string; role: string };
declare const db: { save(u: unknown): void; load(id: number): { name: string } };
declare const auth: { issueToken(u: unknown): { userId: number; isExpired(): boolean } };
declare function add(a: number, b: number): number;
